### ISSUE TITLE
Feature request - Add support for model(:name, Model, explicit: 1, constructed: true)

### ISSUE THREAD

**Issue by adfoster-r7** [sdaubert/rasn1#16]
It'd be great to support passing the explicit/constructed options to the `model` dsl, without needing to set the options on the RASN1::Model directly, i.e.:

```diff
diff --git a/test_case.rb b/test_case.rb
index e0773e06e1..41cbe2803c 100644
--- a/test_case.rb
+++ b/test_case.rb
@@ -72,8 +72,6 @@ end
 #    }
 class EncryptionKey < RASN1::Model
   sequence :EncryptionKey,
-           explicit: 1,
-           constructed: true,
            content: [
              integer(:keytype, explicit: 0, constructed: true),
              octet_string(:keyvalue, explicit: 1, constructed: true),
@@ -100,7 +98,7 @@ class EncTicketPart < RASN1::Model
            content: [
              bit_string(:flags, explicit: 0, constructed: true, bit_length: 32),
-             model(:key, EncryptionKey)
+             model(:key, EncryptionKey, explicit: 1, constructed: true)
            ]
 end
```

Scenario: The Kerberos ASN1 specification defines a reusable `EncryptionKey` model which can be found in multiple places within the ASN1 specification, and I'd like to define the RASN1 model once:

```asn1
   EncKDCRepPart   ::= SEQUENCE {
           key             [0] EncryptionKey,
           # ....
   }

   EncTicketPart   ::= [APPLICATION 3] SEQUENCE {
           # ...
           key                     [1] EncryptionKey,
           # ...
   }

   -- Unencrypted authenticator
   Authenticator   ::= [APPLICATION 2] SEQUENCE  {
           # ...
           subkey                  [6] EncryptionKey OPTIONAL,
   }

   EncAPRepPart    ::= [APPLICATION 27] SEQUENCE {
           # ...
           subkey          [2] EncryptionKey OPTIONAL,
```

This is the current test case:

```ruby
require 'rasn1'

enc_ticket_part_input = (
  "\x63\x82\x03\x2f\x30\x82\x03\x2b\xa0\x07\x03\x05\x00\x50\xa0\x00" \
    "\x00\xa1\x1b\x30\x19\xa0\x03\x02\x01\x17\xa1\x12\x04\x10\x4b\x55" \
    "\x4a\x51\x43\x73\x45\x52\x57\x55\x64\x49\x4b\x48\x6b\x78\xa2\x0a" \
    "\x1b\x08\x44\x57\x2e\x4c\x4f\x43\x41\x4c\xa3\x17\x30\x15\xa0\x03" \
    "\x02\x01\x01\xa1\x0e\x30\x0c\x1b\x0a\x66\x61\x6b\x65\x5f\x6d\x79" \
    "\x73\x71\x6c\xa4\x0b\x30\x09\xa0\x03\x02\x01\x00\xa1\x02\x04\x00" \
    "\xa5\x11\x18\x0f\x32\x30\x32\x32\x30\x36\x32\x30\x31\x30\x33\x39" \
    "\x34\x31\x5a\xa6\x11\x18\x0f\x32\x30\x32\x32\x30\x36\x32\x30\x31" \
    "\x30\x33\x39\x34\x31\x5a\xa7\x11\x18\x0f\x32\x30\x33\x32\x30\x36" \
    "\x31\x37\x31\x30\x33\x39\x34\x31\x5a\xa8\x11\x18\x0f\x32\x30\x33" \
    "\x32\x30\x36\x31\x37\x31\x30\x33\x39\x34\x31\x5a\xaa\x82\x02\x83" \
    "\x30\x82\x02\x7f\x30\x82\x02\x7b\xa0\x03\x02\x01\x01\xa1\x82\x02" \
    "\x72\x04\x82\x02\x6e\x30\x82\x02\x6a\x30\x82\x02\x66\xa0\x04\x02" \
    "\x02\x00\x80\xa1\x82\x02\x5c\x04\x82\x02\x58\x04\x00\x00\x00\x00" \
    "\x00\x00\x00\x01\x00\x00\x00\xbc\x01\x00\x00\x48\x00\x00\x00\x00" \
    "\x00\x00\x00\x0a\x00\x00\x00\x1e\x00\x00\x00\x08\x02\x00\x00\x00" \
    "\x00\x00\x00\x06\x00\x00\x00\x14\x00\x00\x00\x28\x02\x00\x00\x00" \
    "\x00\x00\x00\x07\x00\x00\x00\x14\x00\x00\x00\x40\x02\x00\x00\x00" \
    "\x00\x00\x00\x01\x10\x08\x00\xcc\xcc\xcc\xcc\xac\x01\x00\x00\xcc" \
    "\xcc\xcc\xcc\x6c\x5c\x00\x00\x80\x94\x11\x0c\x92\x84\xd8\x01\xff" \
    "\xff\xff\xff\xff\xff\xff\x7f\xff\xff\xff\xff\xff\xff\xff\x7f\x80" \
    "\x94\x11\x0c\x92\x84\xd8\x01\x00\x00\x00\x00\x00\x00\x00\x00\xff" \
    "\xff\xff\xff\xff\xff\xff\x7f\x14\x00\x14\x00\x97\xdf\x00\x00\x00" \
    "\x00\x00\x00\x60\xe2\x00\x00\x00\x00\x00\x00\x84\x9d\x00\x00\x00" \
    "\x00\x00\x00\x8c\x3b\x00\x00\x00\x00\x00\x00\x94\xc2\x00\x00\x00" \
    "\x00\x00\x00\x7e\xc9\x00\x00\xf4\x01\x00\x00\xf4\x01\x00\x00\x01" \
    "\x02\x00\x00\x05\x00\x00\x00\x72\x34\x00\x00\x00\x00\x00\x00\x00" \
    "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" \
    "\x00\x00\x00\x83\x88\x00\x00\x10\x00\x10\x00\x9c\x94\x00\x00\xa2" \
    "\xd7\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x10\x02\x00\x00\x00" \
    "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" \
    "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" \
    "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x0a" \
    "\x00\x00\x00\x00\x00\x00\x00\x0a\x00\x00\x00\x66\x00\x61\x00\x6b" \
    "\x00\x65\x00\x5f\x00\x6d\x00\x79\x00\x73\x00\x71\x00\x6c\x00\x00" \
    "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" \
    "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" \
    "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" \
    "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x05\x00\x00\x00\x01" \
    "\x02\x00\x00\x07\x00\x00\x00\x00\x02\x00\x00\x07\x00\x00\x00\x08" \
    "\x02\x00\x00\x07\x00\x00\x00\x06\x02\x00\x00\x07\x00\x00\x00\x07" \
    "\x02\x00\x00\x07\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" \
    "\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x08\x00\x00\x00\x44" \
    "\x00\x57\x00\x2e\x00\x4c\x00\x4f\x00\x43\x00\x41\x00\x4c\x00\x05" \
    "\x00\x00\x00\x01\x05\x00\x00\x00\x00\x00\x05\x15\x00\x00\x00\x02" \
    "\x87\x5a\x55\xfd\x66\x7b\xcc\xf7\xbc\xaf\x16\xf4\x01\x00\x00\x00" \
    "\x00\x00\x00\x80\x94\x11\x0c\x92\x84\xd8\x01\x14\x00\x66\x00\x61" \
    "\x00\x6b\x00\x65\x00\x5f\x00\x6d\x00\x79\x00\x73\x00\x71\x00\x6c" \
    "\x00\x00\x00\x76\xff\xff\xff\xe2\x58\xd7\xf5\x06\x17\x64\x9c\x77" \
    "\xc6\xff\x16\x13\x16\x5d\xe3\x00\x00\x00\x00\x76\xff\xff\xff\xaf" \
    "\x64\xea\x1d\x9f\xa0\x63\xe9\xb1\x04\xd2\x53\xc4\x82\xce\x61\x00" \
    "\x00\x00\x00"
).b

#   EncryptionKey   ::= SEQUENCE {
#            keytype         [0] Int32 -- actually encryption type --,
#            keyvalue        [1] OCTET STRING
#    }
class EncryptionKey < RASN1::Model
  sequence :EncryptionKey,
           explicit: 1,
           constructed: true,
           content: [
             integer(:keytype, explicit: 0, constructed: true),
             octet_string(:keyvalue, explicit: 1, constructed: true),
           ]
end

#    -- Encrypted part of ticket
#    EncTicketPart   ::= [APPLICATION 3] SEQUENCE {
#            flags                   [0] TicketFlags,
#            key                     [1] EncryptionKey,
#            crealm                  [2] Realm,
#            cname                   [3] PrincipalName,
#            transited               [4] TransitedEncoding,
#            authtime                [5] KerberosTime,
#            starttime               [6] KerberosTime OPTIONAL,
#            endtime                 [7] KerberosTime,
#            renew-till              [8] KerberosTime OPTIONAL,
#            caddr                   [9] HostAddresses OPTIONAL,
#            authorization-data      [10] AuthorizationData OPTIONAL
#    }
class EncTicketPart < RASN1::Model
  sequence :EncTicketPart,
           class: :application,
           explicit: 3,
           content: [
             bit_string(:flags, explicit: 0, constructed: true, bit_length: 32),
             model(:key, EncryptionKey)
           ]
end

pp EncTicketPart.parse(enc_ticket_part_input)
# (EncTicketPart) EncTicketPart APPLICATION [3] EXPLICIT SEQUENCE:
#   flags CONTEXT [0] EXPLICIT BIT STRING: "P\xA0\x00\x00" (bit length: 32)
#   (EncryptionKey) EncryptionKey CONTEXT [1] EXPLICIT SEQUENCE:
#     keytype CONTEXT [0] EXPLICIT INTEGER: 23
#     keyvalue CONTEXT [1] EXPLICIT OCTET STRING: "KUJQCsERWUdIKHkx"
```

**Comment by adfoster-r7** [comment]
For now I've added a temporary workaround/hack in userspace to support the desired syntax :+1:

For now I went with creating a new class just to see if it would work, but I'll spend some time adding more of the models so there's a real test case to verify against - as I'm not sure if the workaround will work

```diff
diff --git a/test_case.rb b/test_case.rb
index 1dfbeaf2bd..36fc8f6d2a 100644
--- a/test_case.rb
+++ b/test_case.rb
@@ -55,14 +55,39 @@ enc_ticket_part_input = (
     "\x00\x00\x00"
 ).b
 
+# Adding additional methods to RASN1's dsl
+class RASN1::Model
+  def self.custom_model(name, clazz, options={})
+      options.merge!(name: name)
+      proc = proc do |opts|
+          merged_opts = options.merge(opts)
+          root_options = merged_opts.slice(:explicit, :constructed)
+          # XXX: Temporary hack: create a duplicate clazz and set a unique name, i.e. "OldClassName_FieldName_explicit_1_constructed_true"
+          if root_options.any?
+              custom_root_options_clazz_name = "#{clazz.name}_#{name}_#{root_options.map { |k, v| "#{k}_#{v}" }.join('_')}"
+              if !Object.const_defined?(custom_root_options_clazz_name)
+                  new_clazz = Class.new(clazz)
+                  # Assign the name
+                  Object.const_set(custom_root_options_clazz_name, new_clazz)
+                  # Assign the root_options
+                  new_clazz.root_options(root_options)
+              end
+
+              Object.const_get(custom_root_options_clazz_name).new(merged_opts)
+          else
+              clazz.new(merged_opts)
+          end
+      end
+      @root = Elem.new(name, proc, nil)
+  end
+end
+
 #   EncryptionKey   ::= SEQUENCE {
 #            keytype         [0] Int32 -- actually encryption type --,
 #            keyvalue        [1] OCTET STRING
 #    }
 class EncryptionKey < RASN1::Model
   sequence :EncryptionKey,
-           explicit: 1,
-           constructed: true,
            content: [
              integer(:keytype, explicit: 0, constructed: true),
              octet_string(:keyvalue, explicit: 1, constructed: true),
@@ -89,11 +114,26 @@ class EncTicketPart < RASN1::Model
            explicit: 3,
            content: [
              bit_string(:flags, explicit: 0, constructed: true, bit_length: 32),
-             model(:key, EncryptionKey)
+             custom_model(:key, EncryptionKey, explicit: 1, constructed: true),
+             custom_model(:other_fake_key, EncryptionKey, explicit: 2, constructed: 2)
            ]
 end
 
-pp EncTicketPart.parse(enc_ticket_part_input)
+example = EncTicketPart.new(
+  flags:  "P\xA0\x00\x00",
+  key: {
+    keytype: 23,
+    keyvalue: "abc".b
+  },
+  other_fake_key: {
+    keytype: 30,
+    keyvalue: "def".b
+  }
+)
+
+pp EncTicketPart.parse(example.to_der)
+#(EncTicketPart) EncTicketPart APPLICATION [3] EXPLICIT SEQUENCE:
+#  flags CONTEXT [0] EXPLICIT BIT STRING: "P\xA0\x00\x00" (bit length: 32)
+#  (EncryptionKey_key_explicit_1_constructed_true) EncryptionKey CONTEXT [1] EXPLICIT SEQUENCE:
+#    keytype CONTEXT [0] EXPLICIT INTEGER: 23
+#    keyvalue CONTEXT [1] EXPLICIT OCTET STRING: "abc"
+#
+#  (EncryptionKey_other_fake_key_explicit_2_constructed_2) EncryptionKey CONTEXT [2] EXPLICIT SEQUENCE:
+#    keytype CONTEXT [0] EXPLICIT INTEGER: 30
+#    keyvalue CONTEXT [1] EXPLICIT OCTET STRING: "def"
 ```

**Comment by adfoster-r7** [comment]
Thought I'd add an extra datapoint to help describe the use case more - this is the model in csharp:

KDC_REP has a ticket in sequence element 5:

```csharp
    public class KDC_REP : Asn1Sequence
    {
        [Asn1Field(0), Asn1Tag(Asn1TagType.Context, 0)]
        public Asn1Integer pvno { get; set; }
        
        [Asn1Field(1), Asn1Tag(Asn1TagType.Context, 1)]
        public Asn1Integer msg_type { get; set; }
        
        [Asn1Field(2, Optional = true), Asn1Tag(Asn1TagType.Context, 2)]
        public Asn1SequenceOf<PA_DATA> padata { get; set; }
        
        [Asn1Field(3), Asn1Tag(Asn1TagType.Context, 3)]
        public Realm crealm { get; set; }
        
        [Asn1Field(4), Asn1Tag(Asn1TagType.Context, 4)]
        public PrincipalName cname { get; set; }
        
        [Asn1Field(5), Asn1Tag(Asn1TagType.Context, 5)]
        public Ticket ticket { get; set; }
```

https://github.com/microsoft/WindowsProtocolTestSuites/blob/03b3906b9745be72b1852f7ec6ac28ca838029b6/ProtoSDK/KerberosLib/Asn1Code/Rfc4120/KDC_REP.cs#L39-L40

And the Ticket has its own application 1 tag:

```csharp
    [Asn1Tag(Asn1TagType.Application, 1)]
    public class Ticket : Asn1Sequence
    {
```

https://github.com/microsoft/WindowsProtocolTestSuites/blob/03b3906b9745be72b1852f7ec6ac28ca838029b6/ProtoSDK/KerberosLib/Asn1Code/Rfc4120/Ticket.cs#L16-L17

**Comment by sdaubert** [comment]
OK, this is a lot of information. I have to think about it.

I am not fond of `custom_model` as you defined it because it creates new classes.

**Comment by adfoster-r7** [comment]
> I am not fond of custom_model as you defined it because it creates new classes.

Agreed :+1: It was just a quick prototype to see what additional use cases would be required.

I think the final solution would need to support the user specifying an explicit sequence number 5, whilst retaining the original Ticket model's tag class of application 1 - similar to the csharp declarative approach. That support wasn't viable with the class cloning approach, since the two separate pieces of ASN1 metadata need to co-exist

**Comment by sdaubert** [comment]
@adfoster-r7 I've just created a new issue #20 for your `wrapper` request.

**Comment by adfoster-r7** [comment]
Thanks! Do we want to close this issue then, and track the main requirement in your new issue? I think this was just an organically growing ticket that's not as succint as #20 is

**Comment by sdaubert** [comment]
This move tracks the `model(name, klass, options)` feature (with options a hash to declare explicit and/or constructed tags)

**Comment by sdaubert** [comment]
In fact, `model(name, klass, options)` may be done these ways:

```ruby
# First way: use Model.root_options
class Model1 < RASN1::Model
  # define model1
end

class Model2 < Model1
  root_options new_options
end

class Model3 < RASN1::Model
  sequence :seq, content: [model(:a, Model1), model(:b, Model2)]
end

# Second way: use wrapper
class Model1 < RASN1::Model
  # define Model1
end

class Model3 < RASN1::Model
  sequence :seq, content: [model(:a, Model1), wrapper(model(:b, Model1), new_options)]
end
```

@adfoster-r7 is this ok to you?

**Comment by adfoster-r7** [comment]
Thanks! The API change works for me, and it looks like it can parse without throwing an exception now 🎉 

I've progressed a bit further with writing the kerberos model - and it looks like the `to_h` functionality isn't wired up currently - the wrapper doesn't seem to call `to_h` of the `ticket`, similar to the other elements - and it looks like the hash key is `nil` instead of `ticket` for some reason:

```ruby
[2] pry(#<RSpec::ExampleGroups::KerberosModel::KerberosModelAsResponse::Parse>)> described_class.parse(valid_data).to_h
=> {:as_response=>
  {:pvno=>5,
   :msg_type=>11,
   :crealm=>"DEMO.LOCAL",
   :cname=>{:name_type=>1, :name_string=>["juan"]},
   nil=> <-------- Seems to be a nil key
    (Ticket) ticket APPLICATION [1] EXPLICIT SEQUENCE: <--------- Looks like `to_s` instead of `to_h`
  tkt_vno CONTEXT [0] EXPLICIT INTEGER: 5
  realm CONTEXT [1] EXPLICIT KerberosString: "DEMO.LOCAL"
  (PrincipalName_sname_explicit_2_constructed_true) PrincipalName CONTEXT [2] EXPLICIT SEQUENCE:
    name_type CONTEXT [0] EXPLICIT INT32: 1
    name_string CONTEXT [1] EXPLICIT SEQUENCE OF:
      KerberosString: "krbtgt"
      KerberosString: "DEMO.LOCAL"


  (EncryptedData_enc_part_explicit_3_constructed_true) encrypted_data CONTEXT [3] EXPLICIT SEQUENCE:
    etype CONTEXT [0] EXPLICIT INT32: 23
    kvno CONTEXT [1] EXPLICIT UINT32: 2 OPTIONAL
    cipher CONTEXT [2] EXPLICIT OCTET STRING: "U\xE7E\xC3o\xA2(G\xAB\x9C\x86\x13\xEB\x1D\xA8\x98\xECg\x1C\x1F\x15Fk\xE0J\xF2M=\xF7\xE41zO\x15/`\xDD\x98\xA8\xE1\x97ko\xC1$Fl\xA9\x1E\xE26'\xE3\xFA\x99\f\x9Bw\f\xE2X\x02h\xC4T*,]lK\xC8\xBC\x04\x8F\nD'x\xDCK>\x01\xBE\xAC\xF7\x8EzP\xC6>w\xD9e$\xD5\x1A\x18\xA1\x84q\x85\x98/T\x8BV\xE3\xFB,\xE20\x84\x06UU\xEA1\x8B\x84\x00\xE3\x1A\xC3\xA8\xC2\xAC\xC0x?Ght\xCCb\xA6\xCF\xF4k\xAE\xAF'\xDE\x1AM\xB7\xA8\x9Fvzy*B\x12{\xD2\xBE\xC9\x98|D8@\xBDI\xCD>\xDCe\xC7\x8BD\xF5\xA5\xD4f\x0E\xFDX\x9D19'\xD7\xFC\x81\a\xA3*\x1C<"

}}
```

I would have expected `to_h` to give:

```ruby
{
    as_response: {
    pvno: 5,
    msg_type: 11,
    crealm: "DEMO.LOCAL",
    cname: {
      name_type: 1,
      name_string: ["juan"]
    },
    ticket: {
      tkt_vno: 5,
      realm: 'DEMO.local',
      sname: [
      name_type: 1,
      name_string: ['krbtgt', 'DEMO.LOCAL']
      ],
      enc_part: {
      etype: 1337,
      kvno: 1337,
      cipher: "U\xE7E\xC3o\xA2(G\xAB\x9C\x86\x13\xEB\x1D\xA8\x98\xECg\x1C\x1F\x15Fk\xE0J\xF2M=\xF7\xE41zO\x15/`\xDD\x98\xA8\xE1\x97ko\xC1$Fl\xA9\x1E\xE26'\xE3\xFA\x99\f\x9Bw\f\xE2X\x02h\xC4T*,]lK\xC8\xBC\x04\x8F\nD'x\xDCK>\x01\xBE\xAC\xF7\x8EzP\xC6>w\xD9e$\xD5\x1A\x18\xA1\x84q\x85\x98/T\x8BV\xE3\xFB,\xE20\x84\x06UU\xEA1\x8B\x84\x00\xE3\x1A\xC3\xA8\xC2\xAC\xC0x?Ght\xCCb\xA6\xCF\xF4k\xAE\xAF'\xDE\x1AM\xB7\xA8\x9Fvzy*B\x12{\xD2\xBE\xC9\x98|D8@\xBDI\xCD>\xDCe\xC7\x8BD\xF5\xA5\xD4f\x0E\xFDX\x9D19'\xD7\xFC\x81\a\xA3*\x1C<".b
      }
    }
  }
}
```

I've pushed the failing tests here: https://github.com/sdaubert/rasn1/pull/17 in the `describe KerberosModel::AsResponse do` section. Here's the expected rspec diff if my explanation was bad 😄

```ruby
       Diff:
       @@ -2,5 +2,20 @@
        :crealm => "DEMO.LOCAL",
        :msg_type => 11,
        :pvno => 5,
       -:ticket => {:enc_part=>{:cipher=>"U\xE7E\xC3o\xA2(G\xAB\x9C\x86\x13\xEB\x1D\xA8\x98\xECg\x1C\x1F\x15Fk\xE0J\xF2M=\xF7\xE41zO\x15/`\xDD\x98\xA8\xE1\x97ko\xC1$Fl\xA9\x1E\xE26'\xE3\xFA\x99\f\x9Bw\f\xE2X\x02h\xC4T*,]lK\xC8\xBC\x04\x8F\nD'x\xDCK>\x01\xBE\xAC\xF7\x8EzP\xC6>w\xD9e$\xD5\x1A\x18\xA1\x84q\x85\x98/T\x8BV\xE3\xFB,\xE20\x84\x06UU\xEA1\x8B\x84\x00\xE3\x1A\xC3\xA8\xC2\xAC\xC0x?Ght\xCCb\xA6\xCF\xF4k\xAE\xAF'\xDE\x1AM\xB7\xA8\x9Fvzy*B\x12{\xD2\xBE\xC9\x98|D8@\xBDI\xCD>\xDCe\xC7\x8BD\xF5\xA5\xD4f\x0E\xFDX\x9D19'\xD7\xFC\x81\a\xA3*\x1C<", :etype=>1337, :kvno=>1337}, :realm=>"DEMO.local", :sname=>[{:name_string=>["krbtgt", "DEMO.LOCAL"], :name_type=>1}], :tkt_vno=>5},
       +nil => (Ticket) ticket APPLICATION [1] EXPLICIT SEQUENCE:
       +  tkt_vno CONTEXT [0] EXPLICIT INTEGER: 5
       +  realm CONTEXT [1] EXPLICIT KerberosString: "DEMO.LOCAL"
       +  (PrincipalName_sname_explicit_2_constructed_true) PrincipalName CONTEXT [2] EXPLICIT SEQUENCE:
       +    name_type CONTEXT [0] EXPLICIT INT32: 1
       +    name_string CONTEXT [1] EXPLICIT SEQUENCE OF:
       +      KerberosString: "krbtgt"
       +      KerberosString: "DEMO.LOCAL"
       +
       +
       +  (EncryptedData_enc_part_explicit_3_constructed_true) encrypted_data CONTEXT [3] EXPLICIT SEQUENCE:
       +    etype CONTEXT [0] EXPLICIT INT32: 23
       +    kvno CONTEXT [1] EXPLICIT UINT32: 2 OPTIONAL
       +    cipher CONTEXT [2] EXPLICIT OCTET STRING: "U\xE7E\xC3o\xA2(G\xAB\x9C\x86\x13\xEB\x1D\xA8\x98\xECg\x1C\x1F\x15Fk\xE0J\xF2M=\xF7\xE41zO\x15/`\xDD\x98\xA8\xE1\x97ko\xC1$Fl\xA9\x1E\xE26'\xE3\xFA\x99\f\x9Bw\f\xE2X\x02h\xC4T*,]lK\xC8\xBC\x04\x8F\nD'x\xDCK>\x01\xBE\xAC\xF7\x8EzP\xC6>w\xD9e$\xD5\x1A\x18\xA1\x84q\x85\x98/T\x8BV\xE3\xFB,\xE20\x84\x06UU\xEA1\x8B\x84\x00\xE3\x1A\xC3\xA8\xC2\xAC\xC0x?Ght\xCCb\xA6\xCF\xF4k\xAE\xAF'\xDE\x1AM\xB7\xA8\x9Fvzy*B\x12{\xD2\xBE\xC9\x98|D8@\xBDI\xCD>\xDCe\xC7\x8BD\xF5\xA5\xD4f\x0E\xFDX\x9D19'\xD7\xFC\x81\a\xA3*\x1C<"
       +
       +,
```

It's awesome how close the functionality is! 💯

**Comment by sdaubert** [comment]
Fixed in 248ef4a

**Comment by adfoster-r7** [comment]
@sdaubert Thanks! I hit hit another issue when rebasing and trying to remove references to my `custom_model` hack; I think this change is required to preserve the wrapper's intended name:

https://github.com/sdaubert/rasn1/pull/22

**Comment by adfoster-r7** [comment]
Just an update to the above - I've removed all my `custom_model` hacks now, and after #22 is merged in - things are much closer to passing out of the box with the latest code in master 🎉 

I've had to put in one last workaround for a bug I wasn't able to fix here in a nice way over here:
https://github.com/sdaubert/rasn1/pull/17/files#diff-d9dc28d08210c3b2bda21d81fde20522c76237e06cd1ab5a4d8379d2709b123aR597-R600

the `to_h` functionality doesn't seem to use the correct name for the nested child `ticket` sname, I put in a temporary work around for now to get the tests passing when `name` is `nil`:


```ruby
        name = case el
               when Model
                 @elements.key(el)
               when Wrapper
                 @elements.key(el.element)
               else
                 el.name
               end
        [name, private_to_h(el)]

        if name.nil?
          # $stderr.puts "name is nil"
          name = @elements[:ticket].instance_variable_get(:@elements).key(el.element)
        end

        x = [name, private_to_h(el)]
        x
```

I've left the tests failing over in #17 as an example, where the name lookup is failing inside the child `wrapper(model(:ticket, Ticket), explicit: 5, constructed: true),`:

```diff
@@ -1,5 +1,4 @@
-:enc_part => {:cipher=>"U\xE7E\xC3o\xA2(G\xAB\x9C\x86\x13\xEB\x1D\xA8\x98\xECg\x1C\x1F\x15Fk\xE0J\xF2M=\xF7\xE41zO\x15/`\xDD\x98\xA8\xE1\x97ko\xC1$Fl\xA9\x1E\xE26'\xE3\xFA\x99\f\x9Bw\f\xE2X\x02h\xC4T*,]lK\xC8\xBC\x04\x8F\nD'x\xDCK>\x01\xBE\xAC\xF7\x8EzP\xC6>w\xD9e$\xD5\x1A\x18\xA1\x84q\x85\x98/T\x8BV\xE3\xFB,\xE20\x84\x06UU\xEA1\x8B\x84\x00\xE3\x1A\xC3\xA8\xC2\xAC\xC0x?Ght\xCCb\xA6\xCF\xF4k\xAE\xAF'\xDE\x1AM\xB7\xA8\x9Fvzy*B\x12{\xD2\xBE\xC9\x98|D8@\xBDI\xCD>\xDCe\xC7\x8BD\xF5\xA5\xD4f\x0E\xFDX\x9D19'\xD7\xFC\x81\a\xA3*\x1C<", :etype=>23, :kvno=>2},
:realm => "DEMO.LOCAL",
-:sname => {:name_string=>["krbtgt", "DEMO.LOCAL"], :name_type=>1},
:tkt_vno => 5,
+nil => {:cipher=>"U\xE7E\xC3o\xA2(G\xAB\x9C\x86\x13\xEB\x1D\xA8\x98\xECg\x1C\x1F\x15Fk\xE0J\xF2M=\xF7\xE41zO\x15/`\xDD\x98\xA8\xE1\x97ko\xC1$Fl\xA9\x1E\xE26'\xE3\xFA\x99\f\x9Bw\f\xE2X\x02h\xC4T*,]lK\xC8\xBC\x04\x8F\nD'x\xDCK>\x01\xBE\xAC\xF7\x8EzP\xC6>w\xD9e$\xD5\x1A\x18\xA1\x84q\x85\x98/T\x8BV\xE3\xFB,\xE20\x84\x06UU\xEA1\x8B\x84\x00\xE3\x1A\xC3\xA8\xC2\xAC\xC0x?Ght\xCCb\xA6\xCF\xF4k\xAE\xAF'\xDE\x1AM\xB7\xA8\x9Fvzy*B\x12{\xD2\xBE\xC9\x98|D8@\xBDI\xCD>\xDCe\xC7\x8BD\xF5\xA5\xD4f\x0E\xFDX\x9D19'\xD7\xFC\x81\a\xA3*\x1C<", :etype=>23, :kvno=>2},
```

And uncommenting the `name = @elements[:ticket].instance_variable_get(:@elements).key(el.element)` name code line allows the tests to pass again. I'm not sure if I'll be able to come up with a generic solution, I think it might require similar logic to `private_to_h`

**Comment by sdaubert** [comment]
Using wrappers and RASN1 custom type, i don't have your issue to decode a Kerberos ticket. See https://gist.github.com/sdaubert/67ea4550a1cc91c11f1c633d8442955e

**Comment by adfoster-r7** [comment]
I believe the issue only occurs on a nested sequence, so the issue only replicates when the ticket is contained within an ASResponse:

```ruby
{:as_response=>
  {:pvno=>5,
   :msg_type=>11,
   :pa_data=>[],
   :crealm=>"DEMO.LOCAL",
   :cname=>{:name_type=>1, :name_string=>["juan"]},
   :ticket=>
    {:tkt_vno=>5,
     :realm=>"DEMO.LOCAL",
     nil=> <-------------------------------------
      {:etype=>23,
       :kvno=>2,
       :cipher=>
        "U\xE7E\xC3o\xA2(G\xAB\x9C\x86\x13\xEB\x1D\xA8\x98\xECg\x1C\x1F\x15Fk\xE0J\xF2M=\xF7\xE41zO\x15/`\xDD\x98\xA8\xE1\x97ko\xC1$Fl\xA9\x1E\xE26'\xE3\xFA\x99\f\x9Bw\f\xE2X\x02h\xC4T*,]lK\xC8\xBC\x04\x8F\n" +
        "D'x\xDCK>\x01\xBE\xAC\xF7\x8EzP\xC6>w\xD9e$\xD5\x1A\x18\xA1\x84q\x85\x98/T\x8BV\xE3\xFB,\xE20\x84\x06UU\xEA1\x8B\x84\x00\xE3\x1A\xC3\xA8\xC2\xAC\xC0x?Ght\xCCb\xA6\xCF\xF4k\xAE\xAF'\xDE\x1AM\xB7\xA8\x9Fvzy*B\x12{\xD2\xBE\xC9\x98|D8@\xBDI\xCD>\xDCe\xC7\x8BD\xF5\xA5\xD4f\x0E\xFDX\x9D19'\xD7\xFC\x81\a\xA3*\x1C<"}},
   :enc_part=>
    {:etype=>23,
     :kvno=>1,
     :cipher=>
      "\x80\ax6\x16\xC5gc\x83;\xF8Bf\x8A\xEB\xE9\xFE\xBA\x82\x82\x1F=\x15O[p\x19\\\xE2\x02\xD6\xA5\x87S\xBFw#[2)\xDES\xE6\x9D\xE9\xA58qB\xF7\xFE\x93;0O\xBD\x86\xCD+\x88\x84\xB7@v&\xFE\xD69\xA0i\xA3F\x86\xC6\xDF\x1D\xB2\n" +
      "\x91\xC5lUV\xFE\xFB\xCELWtl*\xFC\xBCnz\x19v \xE8a\xC2\x99\xD0\x85s\x99\xC5\x19X^\xAE\xC4\xBB\x14\xAC7\xC9\x80\xAA\xD4\x9D\x12\xC67R\xB7\xF0l+\xED\xFC\xEB\xF6? \xBDQ\x92M\t\xE5\x877\xA7\xFAZ=\x01v\xF3\a\xB1\x98'\xB5\xAC\xB7\x98\xA4\xA6w\xF1\xA3\xF9\xF1\x9FN\x8E\x9Cm\x1AY\f\x9D\xC6;\xA9rSm\xD5\x1D\x00\\*\xE8\xC8\xC4\xB8\x9C\x1F\x05\x9AF\xE7\xA9\xBAx\xA9\xA5\r\x90\xB9\xA8=\x9D\xC2d\xE7\x8C\xE0\xDA\x82\xE8\xD0<\xD6\xEB\xC4\x06\xF5\x19\x9E4rF\x12\xBA\x06Zu\x95\x01{5@2\xA2\\\a!B'\xE0\xDC\x9B\xFCG"}}}
```

Here's a test case using your gist as the starting point:

<details>

<summary>Replication from original gist</summary>

```ruby
require 'rasn1'

RASN1::Types.define_type('Uint32', from: RASN1::Types::Integer) do |value|
  (value >= 0) && (value < 2 ** 32)
end
RASN1::Types.define_type('Int32', from: RASN1::Types::Integer) do |value|
  (value >= -2 ** 31) && (value < 2 ** 31)
end

RASN1::Types.define_type('KerberosString', from: RASN1::Types::IA5String)
RASN1::Types.define_type('KerberosRealm', from: RASN1::Types::KerberosString)

p RASN1::Model.singleton_methods

class PrincipalName < RASN1::Model
  sequence :PrincipalName,
           content: [
             int32(:name_type, explicit: 0, constructed: true),
             sequence_of(:name_string, RASN1::Types::KerberosString, explicit: 1, constructed: true)
           ]
end

class EncryptedData < RASN1::Model
  sequence :encrypted_data,
           content: [
             int32(:etype, explicit: 0, constructed: true),
             uint32(:kvno, explicit: 1, constructed: true, optional: true),
             octet_string(:cipher, explicit: 2, constructed: true)
           ]
end

class PreAuthData < RASN1::Model
  sequence :PreAuthData,
           content: [
             int32(:type, explicit: 1, constructed: true),
             sequence_of(:value, RASN1::Types::KerberosString, explicit: 2, constructed: true)
           ]
end

class Ticket < RASN1::Model
  sequence :ticket,
           class: :application,
           explicit: 1,
           content: [
             integer(:tkt_vno, explicit: 0, constructed: true, default: 5),
             kerberosrealm(:realm, explicit: 1, constructed: true),
             wrapper(model(:sname, PrincipalName), explicit: 2, constructed: true),
             wrapper(model(:enc_part, EncryptedData), explicit: 3, constructed: true)
           ]
end

class AsResponse < RASN1::Model
  sequence :as_response,
           class: :application,
           explicit: 11,
           content: [
             integer(:pvno, explicit: 0, constructed: true),
             # 11 AS, 13 TGS
             integer(:msg_type, explicit: 1, constructed: true),
             sequence_of(:pa_data, PreAuthData, explicit: 2, constructed: true, optional: true),
             kerberosrealm(:crealm, explicit: 3, constructed: true),
             wrapper(model(:cname, PrincipalName), explicit: 4, constructed: true),
             wrapper(model(:ticket, Ticket), explicit: 5, constructed: true),
             wrapper(model(:enc_part, EncryptedData), explicit: 6, constructed: true)
           ]
end

as_response = AsResponse.new(
  {
    pvno: 5,
    msg_type: 11,
    crealm: "DEMO.LOCAL",
    cname: {
      name_type: 1,
      name_string: ["juan"]
    },
    ticket: {
      tkt_vno: 5,
      realm: 'DEMO.LOCAL',
      sname: {
        name_type: 1,
        name_string: ['krbtgt', 'DEMO.LOCAL']
      },
      enc_part: {
        etype: 23,
        kvno: 2,
        cipher: "U\xE7E\xC3o\xA2(G\xAB\x9C\x86\x13\xEB\x1D\xA8\x98\xECg\x1C\x1F\x15Fk\xE0J\xF2M=\xF7\xE41zO\x15/`\xDD\x98\xA8\xE1\x97ko\xC1$Fl\xA9\x1E\xE26'\xE3\xFA\x99\f\x9Bw\f\xE2X\x02h\xC4T*,]lK\xC8\xBC\x04\x8F\nD'x\xDCK>\x01\xBE\xAC\xF7\x8EzP\xC6>w\xD9e$\xD5\x1A\x18\xA1\x84q\x85\x98/T\x8BV\xE3\xFB,\xE20\x84\x06UU\xEA1\x8B\x84\x00\xE3\x1A\xC3\xA8\xC2\xAC\xC0x?Ght\xCCb\xA6\xCF\xF4k\xAE\xAF'\xDE\x1AM\xB7\xA8\x9Fvzy*B\x12{\xD2\xBE\xC9\x98|D8@\xBDI\xCD>\xDCe\xC7\x8BD\xF5\xA5\xD4f\x0E\xFDX\x9D19'\xD7\xFC\x81\a\xA3*\x1C<".b
      }
    },
    enc_part: {
      etype: 23,
      kvno: 1,
      cipher: "\x80\ax6\x16\xC5gc\x83;\xF8Bf\x8A\xEB\xE9\xFE\xBA\x82\x82\x1F=\x15O[p\x19\\\xE2\x02\xD6\xA5\x87S\xBFw#[2)\xDES\xE6\x9D\xE9\xA58qB\xF7\xFE\x93;0O\xBD\x86\xCD+\x88\x84\xB7@v&\xFE\xD69\xA0i\xA3F\x86\xC6\xDF\x1D\xB2\n\x91\xC5lUV\xFE\xFB\xCELWtl*\xFC\xBCnz\x19v \xE8a\xC2\x99\xD0\x85s\x99\xC5\x19X^\xAE\xC4\xBB\x14\xAC7\xC9\x80\xAA\xD4\x9D\x12\xC67R\xB7\xF0l+\xED\xFC\xEB\xF6? \xBDQ\x92M\t\xE5\x877\xA7\xFAZ=\x01v\xF3\a\xB1\x98'\xB5\xAC\xB7\x98\xA4\xA6w\xF1\xA3\xF9\xF1\x9FN\x8E\x9Cm\x1AY\f\x9D\xC6;\xA9rSm\xD5\x1D\x00\\*\xE8\xC8\xC4\xB8\x9C\x1F\x05\x9AF\xE7\xA9\xBAx\xA9\xA5\r\x90\xB9\xA8=\x9D\xC2d\xE7\x8C\xE0\xDA\x82\xE8\xD0<\xD6\xEB\xC4\x06\xF5\x19\x9E4rF\x12\xBA\x06Zu\x95\x01{5@2\xA2\\\a!B'\xE0\xDC\x9B\xFCG".b
    }
  }
)

p as_response
pp as_response.to_h
```
</details>

I believe the name calculation should be updated to reference the sequence's parent, rather than assuming it's the name can be looked up within `@elements` - which is why my hack works here:

https://github.com/sdaubert/rasn1/commit/2e8db089325e471e8139c4357911804a2a6cecf9

**Comment by adfoster-r7** [comment]
I created a smaller test case for the bug:

```diff
diff --git a/spec/model_spec.rb b/spec/model_spec.rb
index 5bfee99..8f57d80 100644
--- a/spec/model_spec.rb
+++ b/spec/model_spec.rb
@@ -239,6 +239,13 @@ module RASN1 # rubocop:disable Metrics/moduleLength
         model[:a_record][:house] = 5
         expect(model.to_h).to eq({ seq: { a_record: { id: 4, house: 5 } } })
       end
+
+      it 'generates a Hash image of a model with a nested wrapped submodel' do
+        model = ModelWithNestedExplicitModel.new
+        model[:body][:a_record][:id] = 4
+        model[:body][:a_record][:house] = 5
+        expect(model.to_h).to eq({ model_with_nested_explicit_model: { body: { a_record: { id: 4, house: 5 } } } })
+      end
     end
 
     describe '#to_der' do
diff --git a/spec/spec_helper.rb b/spec/spec_helper.rb
index 5e34e08..442644d 100644
--- a/spec/spec_helper.rb
+++ b/spec/spec_helper.rb
@@ -80,4 +80,11 @@ module TestModel
     sequence :seq,
              content: [wrapper(model(:a_record, ModelTest3), explicit: 6)]
   end
+
+  class ModelWithNestedExplicitModel < RASN1::Model
+    sequence :model_with_nested_explicit_model,
+             content: [
+               model(:body, ModelWithExplicitWrapper)
+             ]
+  end
 end

```

Fails with:
```
Failures:

  1) RASN1::Model#to_h generates a Hash image of a model with a nested wrapped submodel
     Failure/Error: expect(model.to_h).to eq({ model_with_nested_explicit_model: { body: { a_record: { id: 4, house: 5 } } } })
     
       expected: {:model_with_nested_explicit_model=>{:body=>{:a_record=>{:house=>5, :id=>4}}}}
            got: {:model_with_nested_explicit_model=>{:body=>{nil=>{:house=>5, :id=>4}}}}
     
       (compared using ==)
     
       Diff:
       @@ -1 +1 @@
       -:model_with_nested_explicit_model => {:body=>{:a_record=>{:house=>5, :id=>4}}},
       +:model_with_nested_explicit_model => {:body=>{nil=>{:house=>5, :id=>4}}},
       
     # ./spec/model_spec.rb:247:in `block (3 levels) in <module:RASN1>'

```

And I can get the tests passing with:

```
diff --git a/lib/rasn1/model.rb b/lib/rasn1/model.rb
index 95bad6a..dbb7160 100644
--- a/lib/rasn1/model.rb
+++ b/lib/rasn1/model.rb
@@ -593,6 +593,9 @@ module RASN1
                else
                  el.name
                end
+        if name.nil?
+          name = @elements[:body].instance_variable_get(:@elements).key(el.element)
+        end
         [name, private_to_h(el)]
       end
       ary.compact!
```

I'm thinking it might be possible to fix this properly by passing in the sequence parent as `to_h` traverses its children

**Comment by sdaubert** [comment]
Fix done in Model.
I also changed way explicit models give acces to their inner element.

**Comment by adfoster-r7** [comment]
Awesome! That's a really elegant solution 🎉 

I've been able to convert over a lot more models now, thanks! I hit one edgecase with the `optional` functionality which doesn't work with the new wrapper functionality, i.e.

```ruby
wrapper(model(:authorization_data, AuthorizationData), explicit: 8, constructed: true, optional: true)
```

<details>

<summary>full replication from original gist</summary>

```ruby
require 'rasn1'

RASN1::Types.define_type('Uint32', from: RASN1::Types::Integer) do |value|
  (value >= 0) && (value < 2 ** 32)
end
RASN1::Types.define_type('Int32', from: RASN1::Types::Integer) do |value|
  (value >= -2 ** 31) && (value < 2 ** 31)
end
RASN1::Types.define_type('KerberosMicroseconds', from: RASN1::Types::Integer)
RASN1::Types.define_type('KerberosTime', from: RASN1::Types::GeneralizedTime)
RASN1::Types.define_type('KerberosFlags', from: RASN1::Types::BitString)
RASN1::Types.define_type('KerberosTicketFlags', from: RASN1::Types::KerberosFlags)

class KerberosString < RASN1::Types::IA5String
  # Important: Using GraphicString's id value of 27, instead of IA5String's ID of 22
  ID = 27

  # Get ASN.1 type
  # @return [String]
  def self.type
    'KerberosString'
  end
end

RASN1::Types.define_type('KerberosString', from: KerberosString)
RASN1::Types.define_type('KerberosRealm', from: RASN1::Types::KerberosString)

p RASN1::Model.singleton_methods

class PrincipalName < RASN1::Model
  sequence :PrincipalName,
           content: [
             int32(:name_type, explicit: 0, constructed: true),
             sequence_of(:name_string, RASN1::Types::KerberosString, explicit: 1, constructed: true)
           ]
end

class HostAddress < RASN1::Model
  sequence :host_address,
           content: [
             int32(:addr_type, explicit: 0, constructed: true),
             octet_string(:address, explicit: 1, constructed: true)
           ]
end

class HostAddresses < RASN1::Model
  sequence_of :addresses, HostAddress
end

class AuthorizationDataItem < RASN1::Model
  sequence :authroization_data_item,
           content: [
             int32(:ad_type, explicit: 0, constructed: true),
             octet_string(:ad_data, explicit: 1, constructed: true),
           ]
end

class AuthorizationData < RASN1::Model
  sequence_of :items, AuthorizationDataItem, constructed: true
end

class EncryptionKey < RASN1::Model
  sequence :EncryptionKey,
           content: [
             int32(:keytype, explicit: 0, constructed: true),
             octet_string(:keyvalue, explicit: 1, constructed: true),
           ]
end

class Checksum < RASN1::Model
  sequence :Checksum,
           content: [
             int32(:cksumtype, explicit: 0, constructed: true),
             octet_string(:checksum, explicit: 1, constructed: true)
           ]
end

class Authenticator < RASN1::Model
  sequence :Authenticator,
           class: :application,
           explicit: 2,
           content: [
             integer(:vno, explicit: 0, constructed: true),
             kerberos_realm(:crealm, explicit: 1, constructed: true),
             wrapper(model(:cname, PrincipalName), explicit: 2, constructed: true),
             wrapper(model(:cksum, Checksum), explicit: 3, constructed: true, optional: true),
             kerberos_microseconds(:cusec, explicit: 4, constructed: true),
             kerberos_time(:ctime, explicit: 5, constructed: true, optional: true),
             wrapper(model(:subkey, EncryptionKey), explicit: 6, constructed: true, optional: true),
             uint32(:seq_number, explicit: 7, constructed: true, optional: true),
             # vvvvvvvvvvvvvvvvvvvvvvvvvv here vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
             wrapper(model(:authorization_data, AuthorizationData), explicit: 8, constructed: true, optional: true)
           ]
end

valid_data = (
  "\x62\x7c\x30\x7a\xa0\x03\x02\x01\x05\xa1\x0c\x1b\x0a\x44\x45\x4d\x4f\x2e\x4c\x4f" \
  "\x43\x41\x4c\xa2\x11\x30\x0f\xa0\x03\x02\x01\x01\xa1\x08\x30\x06\x1b\x04\x6a\x75" \
  "\x61\x6e\xa3\x1b\x30\x19\xa0\x03\x02\x01\x07\xa1\x12\x04\x10\x9e\xf0\x84\xd6\x81" \
  "\xe5\x16\x02\x32\xb1\xc3\x4e\xad\x83\x1d\x43\xa4\x05\x02\x03\x0a\xf8\x98\xa5\x11" \
  "\x18\x0f\x32\x30\x31\x34\x31\x32\x31\x36\x32\x32\x35\x30\x34\x36\x5a\xa6\x1b\x30" \
  "\x19\xa0\x03\x02\x01\x17\xa1\x12\x04\x10\x7d\x63\xdd\x79\x73\x67\xce\x86\xbb\x5f" \
  "\x2b\x8a\xba\x58\xfd\x6e"
).b

# expected = {
#   Authenticator: {
#     vno: 5,
#     crealm: "DEMO.LOCAL",
#     cname: {
#       name_type: 1,
#       name_string: ["juan"]
#     },
#     cksum: {
#       cksumtype: 7,
#       checksum: "\x9E\xF0\x84\xD6\x81\xE5\x16\x022\xB1\xC3N\xAD\x83\x1DC".b
#     },
#     cusec: 719000,
#     ctime: DateTime.parse('2014-12-16T22:50:46+00:00'),
#     subkey: {
#       keytype: 23,
#       keyvalue: "}c\xDDysg\xCE\x86\xBB_+\x8A\xBAX\xFDn".b
#     },
#     authorization_data: []
#   }
# }

authenticator = Authenticator.parse(valid_data)

p authenticator
pp authenticator.to_h

if valid_data != authenticator.to_der
  raise 'failure - input der should match to_der result'
else
  puts 'success!'
end
```

Which fails during parsing:

```
Traceback (most recent call last):
        9: from ./test_case.rb:127:in `<main>'
        8: from /Users/user/Documents/code/rasn1/lib/rasn1/model.rb:215:in `parse'
        7: from /Users/user/Documents/code/rasn1/lib/rasn1/model.rb:385:in `parse!'
        6: from /Users/user/Documents/code/rasn1/lib/rasn1/types/base.rb:226:in `parse!'
        5: from /Users/user/Documents/code/rasn1/lib/rasn1/types/base.rb:229:in `parse!'
        4: from /Users/user/Documents/code/rasn1/lib/rasn1/types/sequence.rb:82:in `der_to_value'
        3: from /Users/user/Documents/code/rasn1/lib/rasn1/types/sequence.rb:82:in `each'
        2: from /Users/user/Documents/code/rasn1/lib/rasn1/types/sequence.rb:83:in `block in der_to_value'
        1: from /Users/user/Documents/code/rasn1/lib/rasn1/wrapper.rb:102:in `parse!'
/Users/user/Documents/code/rasn1/lib/rasn1/model.rb:385:in `parse!': undefined method `force_encoding' for nil:NilClass (NoMethodError)
```

I patched that with:
```
diff --git a/lib/rasn1/wrapper.rb b/lib/rasn1/wrapper.rb
index b86908b..901ad44 100644
--- a/lib/rasn1/wrapper.rb
+++ b/lib/rasn1/wrapper.rb
@@ -44,6 +44,7 @@ module RASN1
       opts = options.dup
       @explicit = opts.delete(:explicit)
       @implicit = opts.delete(:implicit)
+      @optional = opts.delete(:optional)
       if explicit?
         @options = opts
         # ExplicitWrapper is a hand-made explicit tag, but we have to use its implicit option
@@ -72,6 +73,12 @@ module RASN1
       !!@implicit
     end
 
+    # Say if wrapper is an optional one
+    # @return [Boolean]
+    def optional?
+      !!@optional
+    end
+

    # Parse a DER string. This method updates object.
    # @param [String] der DER string
    # @param [Boolean] ber if +true+, accept BER encoding
    # @return [Integer] total number of parsed bytes
    # @raise [ASN1Error] error on parsing
    def parse!(der, ber: false)
+      if der.empty?
+        return 0 if optional?
+
+        raise ASN1Error, 'Expected ANY but get nothing'
+      end
```

But it looks like the `to_der` functionality outputs the extra element with the wrapper too - this is the `diff` between the input der between `valid_data != authenticator.to_der`:


```diff
 Application 2 (1 elem)
-   SEQUENCE (7 elem)
+   SEQUENCE (8 elem)
     [0] (1 elem)
       INTEGER 5
     [1] (1 elem)
       GeneralString DEMO.LOCAL
     [2] (1 elem)
       SEQUENCE (2 elem)
         [0] (1 elem)
           INTEGER 1
         [1] (1 elem)
           SEQUENCE (1 elem)
             GeneralString juan
     [3] (1 elem)
       SEQUENCE (2 elem)
         [0] (1 elem)
           INTEGER 7
         [1] (1 elem)
           OCTET STRING (16 byte) 9EF084D681E5160232B1C34EAD831D43
     [4] (1 elem)
       INTEGER 719000
     [5] (1 elem)
       GeneralizedTime 2014-12-16 22:50:46 UTC
     [6] (1 elem)
       SEQUENCE (2 elem)
         [0] (1 elem)
           INTEGER 23
         [1] (1 elem)
           OCTET STRING (16 byte) 7D63DD797367CE86BB5F2B8ABA58FD6E
+    [8] (1 elem)
+      SEQUENCE (0 elem)
```

My workaround to get the final `valid_data != authenticator.to_der` test passing was this:

```
diff --git a/lib/rasn1/wrapper.rb b/lib/rasn1/wrapper.rb
index b86908b..e5759fe 100644
--- a/lib/rasn1/wrapper.rb
+++ b/lib/rasn1/wrapper.rb
     # Convert wrapper and its element to a DER string
     # @return [String]
     def to_der
+      # TODO: Fix optional sequence wrapper
+      if optional? && element.is_a?(AuthorizationData) && element.value.empty?
+        return ''.b
+      end
+
       if implicit?
         el = generate_implicit_element
         el.to_der
```

I was thinking the real solution might need something similar to the `Any` element logic maybe? 🤔 

```ruby
      def to_der
        if value?
          case @value
          when Base, Model
            @value.to_der
          else
            @value.to_s
          end
        else
          optional? ? '' : Null.new.to_der
        end
      end
```

**Comment by sdaubert** [comment]
Should be fixed now (34e91b8b)

**Comment by adfoster-r7** [comment]
Thanks! With the latest change I've been able to successfully the critical parts of the kerberos spec now 🎉 

I'm hitting some edgecases with `.to_der` not outputting the same input - similar to the above `valid_data != authenticator.to_der` scenario.

I've an example of that here:

```ruby
  class KrbError < KerberosModel
    sequence sequence_name,
             class: :application,
             explicit: 30,
             content: [
               integer(:pvno, explicit: 0, constructed: true),
               integer(:msg_type, explicit: 1, constructed: true),
               kerberos_time(:ctime, explicit: 2, constructed: true, optional: true),
               kerberos_microseconds(:cusec, explicit: 3, constructed: true, optional: true),
               kerberos_time(:stime, explicit: 4, constructed: true),
               kerberos_microseconds(:susec, explicit: 5, constructed: true),
               kerberos_uint32(:error_code, explicit: 6, constructed: true),
               kerberos_realm(:crealm, explicit: 7, constructed: true, optional: true),
               wrapper(model(:cname, PrincipalName), explicit: 8, constructed: true, optional: true), # <--------
               kerberos_realm(:realm, explicit: 9, constructed: true),
               wrapper(model(:sname, PrincipalName), explicit: 10, constructed: true),
               kerberos_string(:e_text, explicit: 11, constructed: true, optional: true),
               octet_string(:e_data, explicit: 12, constructed: true, optional: true),
             ]
  end
```

Here's an example of the input der diffed with the rasn1 `.to_der` call (using a diff against the output of https://lapo.it/asn1js)

```diff
Application 30 (1 elem)
-  SEQUENCE (7 elem)
+  SEQUENCE (8 elem)
    [0] (1 elem)
      INTEGER 5
    [1] (1 elem)
      INTEGER 30
    [4] (1 elem)
      GeneralizedTime 2014-12-14 06:54:01 UTC
    [5] (1 elem)
      INTEGER 724315
    [6] (1 elem)
      INTEGER 60
+    [8] (1 elem)
+      SEQUENCE (2 elem)
+        [0] (1 elem)
+          INTEGER 0
+        [1] (1 elem)
+          SEQUENCE (0 elem)
    [9] (1 elem)
      GeneralString DEMO.LOCAL
    [10] (1 elem)
      SEQUENCE (2 elem)
        [0] (1 elem)
          INTEGER 2
        [1] (1 elem)
          SEQUENCE (2 elem)
            GeneralString krbtgt
            GeneralString DEMO.LOCAL
```

I've marked those tests now as pending over in https://github.com/sdaubert/rasn1/pull/17 :

```ruby
    # TODO: Fails due to the optional fields
    it_behaves_like 'a model that produces the same binary data when to_der is called', pending: true
```

Should I try to make minimal test cases out of those scenarios to help, or would that branch work as

**Comment by sdaubert** [comment]
@adfoster-r7 yes please make minimal testcase, and also please open another issue, as the subject is not the same. I will close this one.

**Comment by adfoster-r7** [comment]
Will do; Thanks!

**Comment by adfoster-r7** [comment]
Created a separate issue now, and I've got a more minified example now - https://github.com/sdaubert/rasn1/issues/24#issuecomment-1304582356
